server/src/test/java/org/elasticsearch/snapshots/SnapshotResiliencyTests.java

```diff
@@ -197,12 +233,62 @@ public void testSuccessfulSnapshot() {
+                                                        new SearchRequest(index).source(
+                                                            new SearchSourceBuilder().size(0).trackTotalHits(true)
+                                                        ),
+                                                        assertNoFailureListener(r -> {
+                                                            assertEquals(
+                                                                (long) documents,
+                                                                Objects.requireNonNull(r.getHits().getTotalHits()).value
+                                                            );
+                                                            documentCountVerified.set(true);
+                                                        }));
+                                                })
+                                            )));
+                                    }));
+                                final AtomicInteger countdown = new AtomicInteger(documents);
+                                masterNode.client.admin().indices().putMapping(
+                                    new PutMappingRequest(index).type("_doc").source("foo", "type=text"),
+                                    assertNoFailureListener(r -> {
+                                            for (int i = 0; i < documents; ++i) {
+                                                masterNode.client.bulk(
+                                                    new BulkRequest().add(new IndexRequest(index).source(
+                                                        Collections.singletonMap("foo", "bar" + i)))
+                                                        .setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE),
+                                                    assertNoFailureListener(
+                                                        bulkResponse -> {
+                                                            assertFalse(
+                                                                "Failures in bulkresponse: " + bulkResponse.buildFailureMessage(),
+                                                                bulkResponse.hasFailures());
+                                                            if (countdown.decrementAndGet() == 0) {
+                                                                afterIndexing.run();
+                                                            }
+                                                        }));
+                                            }
+                                            if (documents == 0) {
+                                                afterIndexing.run();
+                                            }
+                                        }
+                                    ));
+                            }))));
+        runUntil(
```

Comment:
Why do we need 3 different atomic variables? Is not it enough just to have documentCountVerified?